File: external/js/cothority/lib/omniledger/Invoke.js
const Argument = require("./Argument");

/**
 * Invoke is an operation that an Instruction can take, it should be used
 * for mutating an object.
 */
class Invoke {
  /**
   * Creates an new invoke action
   *
   * @param {string} command - the command to invoke in the contract
   * @param {Array<Argument>} argumentsList - the arguments for the contract
   */
  constructor(command, argumentsList) {
    this._command = command;
    this._arguments = argumentsList;
  }

  /**
   * Creates a new Invoke command from one Argument
   *
   * @param {string} command - the command to invoke in the contract
   * @param {string} argument - name of the argument
   * @param {Uint8Array} value - the value of the argument
   * @return {Invoke}
   */
  static fromArgumentInfo(command, argument, value) {
    return new Invoke(command, [new Argument(argument, value)]);
  }

  /**
   * Getter for the command
   *
   * @return {string}
   */
  get command() {
    return this._command;
  }

  /**
   * Getter for the arguments
   *
   * @return {Array<Argument>}
   */
  get argumentsList() {
    return this._arguments;
  }

  /**
   * Create an object with all the necessary field needed to be a valid message
   * in the sense of protobufjs. This object can then be used with the "create"
   * method of protobuf
   *
   * @return {Object}
   */
  toProtobufValidMessage() {
    return {
      command: this._command,
      args: this.argumentsList.map(arg => arg.toProtobufValidMessage())
    };
  }
}

module.exports = Invoke;

File: external/java/src/main/java/ch/epfl/dedis/lib/byzcoin/Invoke.java
package ch.epfl.dedis.lib.byzcoin;

import ch.epfl.dedis.proto.ByzCoinProto;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Invoke is an operation that an Instruction can take, it should be used for mutating an object.
 */
public class Invoke {
    private String command;
    private List<Argument> arguments;

    /**
     * Constructor for the invoke action.
     * @param command The command to invoke in the contract.
     * @param arguments The arguments for the contract.
     */
    public Invoke(String command, List<Argument> arguments) {
        this.command = command;
        this.arguments = arguments;
    }

    /**
     * Constructor from one name/value.
     * @param command
     * @param name
     * @param value
     */
    public Invoke(String command, String name, byte[] value){
        this(command, Arrays.asList(new Argument(name, value)));
    }

    /**
     * Constructo from protobuf.
     * @param proto
     */
    public Invoke(ByzCoinProto.Invoke proto) {
        command = proto.getCommand();
        arguments = new ArrayList<Argument>();
        for (ByzCoinProto.Argument a : proto.getArgsList()) {
            arguments.add(new Argument(a));
        }
    }

    /**
     * Getter for the command.
     * @return The command.
     */
    public String getCommand() {
        return command;
    }

    /**
     * Getter for the arguments
     * @return The arguments.
     */
    public List<Argument> getArguments() {
        return arguments;
    }

    /**
     * Converts this object to the protobuf representation.
     * @return The protobuf representation.
     */
    public ByzCoinProto.Invoke toProto() {
        ByzCoinProto.Invoke.Builder b = ByzCoinProto.Invoke.newBuilder();
        b.setCommand(this.command);
        for (Argument a : this.arguments) {
            b.addArgs(a.toProto());
        }
        return b.build();
    }
}
